File: tests/fixtures/python/class_methods.py
"""
Class with methods for testing Python method chunking
"""


class Calculator:
    """Calculator class with basic arithmetic operations."""

    def __init__(self, precision: int = 2):
        """Constructor initializes the calculator.

        Args:
            precision: Number of decimal places (default: 2)
        """
        self.precision = precision
        self.history = []

    def add(self, a: float, b: float) -> float:
        """Adds two numbers.

        Args:
            a: First number
            b: Second number

        Returns:
            Sum of a and b
        """
        result = a + b
        self.history.append(f"{a} + {b} = {result}")
        return self._round_to_precision(result)

    def subtract(self, a: float, b: float) -> float:
        """Subtracts b from a.

        Args:
            a: First number
            b: Second number

        Returns:
            Difference of a and b
        """
        result = a - b
        self.history.append(f"{a} - {b} = {result}")
        return self._round_to_precision(result)

    def multiply(self, a: float, b: float) -> float:
        """Multiplies two numbers."""
        result = a * b
        return self._round_to_precision(result)

    async def fetch_calculation(self, expression: str) -> float:
        """Async method that fetches calculation result from API.

        Args:
            expression: Mathematical expression to evaluate

        Returns:
            Calculated value
        """
        import aiohttp
        async with aiohttp.ClientSession() as session:
            url = f"/api/calculate?expr={expression}"
            async with session.get(url) as response:
                data = await response.json()
                return data['value']

    def calculate_steps(self, a: float, b: float):
        """Generator method that yields calculation steps.

        Args:
            a: First number
            b: Second number

        Yields:
            Calculation step descriptions
        """
        yield f"Starting calculation: {a} + {b}"
        result = a + b
        yield f"Intermediate result: {result}"
        yield f"Final result: {self._round_to_precision(result)}"

    def _round_to_precision(self, value: float) -> float:
        """Private helper method to round to precision.

        Args:
            value: Value to round

        Returns:
            Rounded value
        """
        return round(value, self.precision)

    def get_history(self) -> list[str]:
        """Gets the calculation history.

        Returns:
            List of calculation history strings
        """
        return self.history.copy()

    def clear_history(self) -> None:
        """Clears the history."""
        self.history.clear()

    @property
    def total_calculations(self) -> int:
        """Property that returns total number of calculations.

        Returns:
            Number of calculations in history
        """
        return len(self.history)

    @total_calculations.setter
    def total_calculations(self, value: int) -> None:
        """Setter for total_calculations (not really useful, just for testing)."""
        pass  # Can't actually set this


class ScientificCalculator(Calculator):
    """Extended calculator with scientific operations.

    Inherits from Calculator and adds scientific functions.
    """

    def __init__(self):
        """Constructor with default high precision."""
        super().__init__(precision=10)

    def power(self, base: float, exponent: float) -> float:
        """Calculates power.

        Args:
            base: Base number
            exponent: Exponent

        Returns:
            base^exponent
        """
        return base ** exponent

    def sqrt(self, value: float) -> float:
        """Calculates square root.

        Args:
            value: Number to take square root of

        Returns:
            Square root of value
        """
        import math
        return math.sqrt(value)


# Multiple inheritance example
class LoggingMixin:
    """Mixin that adds logging capability."""

    def log(self, message: str) -> None:
        """Logs a message.

        Args:
            message: Message to log
        """
        print(f"[LOG] {message}")


class LoggingCalculator(Calculator, LoggingMixin):
    """Calculator with logging capability."""

    def add(self, a: float, b: float) -> float:
        """Adds two numbers with logging."""
        self.log(f"Adding {a} + {b}")
        return super().add(a, b)

File: tests/fixtures/javascript/class-methods.js
/**
 * Class with methods for testing JavaScript method chunking
 */

/**
 * Calculator class with basic arithmetic operations
 */
class Calculator {
  /**
   * Constructor initializes the calculator
   * @param {number} precision Number of decimal places
   */
  constructor(precision = 2) {
    this.precision = precision;
    this.history = [];
  }

  /**
   * Adds two numbers
   * @param {number} a First number
   * @param {number} b Second number
   * @returns {number} Sum of a and b
   */
  add(a, b) {
    const result = a + b;
    this.history.push(`${a} + ${b} = ${result}`);
    return this.roundToPrecision(result);
  }

  /**
   * Subtracts b from a
   */
  subtract(a, b) {
    const result = a - b;
    this.history.push(`${a} - ${b} = ${result}`);
    return this.roundToPrecision(result);
  }

  /**
   * Async method that fetches calculation result from API
   */
  async fetchCalculation(expression) {
    const response = await fetch(`/api/calculate?expr=${expression}`);
    const result = await response.json();
    return result.value;
  }

  /**
   * Generator method that yields calculation steps
   */
  *calculateSteps(a, b) {
    yield `Starting calculation: ${a} + ${b}`;
    const result = a + b;
    yield `Intermediate result: ${result}`;
    yield `Final result: ${this.roundToPrecision(result)}`;
  }

  /**
   * Private helper method (using # private fields - ES2022)
   */
  roundToPrecision(value) {
    const multiplier = Math.pow(10, this.precision);
    return Math.round(value * multiplier) / multiplier;
  }

  /**
   * Gets the calculation history
   */
  getHistory() {
    return [...this.history];
  }
}

/**
 * Extended calculator with scientific operations
 */
class ScientificCalculator extends Calculator {
  /**
   * Constructor with default high precision
   */
  constructor() {
    super(10);
  }

  /**
   * Calculates power
   */
  power(base, exponent) {
    return Math.pow(base, exponent);
  }

  /**
   * Calculates square root
   */
  sqrt(value) {
    return Math.sqrt(value);
  }
}

module.exports = { Calculator, ScientificCalculator };
